src/v/datalake/translation/scheduling.cc

```diff
@@ -0,0 +1,428 @@
+          units);
+        return static_cast<size_t>(units) < _reservation_block_size;
+    }
+
+    ss::future<reservation> reserve_memory(ss::abort_source& as) override {
+        auto nr = _reservation_block_size;
+        // fast path
+        auto opt_units = ss::try_get_units(_available_memory, nr);
+        if (opt_units) {
+            co_return std::move(opt_units.value());
+        }
+        _notifier.notify_memory_exhausted();
+        co_return co_await ss::get_units(_available_memory, nr, as);
+    }
+
+    size_t allocated_memory() const override {
+        return _total_memory - _available_memory.available_units();
+    }
+
+private:
+    const size_t _total_memory;
+    // note: the semaphore should be alive until all the reserved units are
+    // deposited back.
+    ssx::semaphore _available_memory;
+    const size_t _reservation_block_size;
+    scheduling_notifications& _notifier;
+};
+
+std::unique_ptr<reservations_tracker> make_default(
+  size_t total_memory,
+  size_t memory_block_size,
+  scheduling_notifications& notifier) {
+    return std::make_unique<default_reservations_tracker>(
+      total_memory, memory_block_size, notifier);
+}
+
+std::ostream& operator<<(std::ostream& os, const translation_status& status) {
+    fmt::print(
+      os,
+      "{{target_lag: {}, next_checkpoint: {}, memory_reserved: {}}}",
+      std::chrono::duration_cast<std::chrono::milliseconds>(status.target_lag),
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+        status.next_checkpoint_deadline - clock::now()),
+      status.memory_bytes_reserved);
+    return os;
+}
+
+std::ostream& operator<<(std::ostream& os, const translator& t) {
+    fmt::print(os, "{{id: {}, status: {}}}", t.id(), t.status());
+    return os;
+}
+
+translator_executable::translator_executable(
+  translator_executable&& other) noexcept {
+    if (this != &other) {
+        *this = std::move(other);
+    }
+}
+
+translator_executable&
+translator_executable::operator=(translator_executable&& other) noexcept {
+    _waiting_hook.swap_nodes(other._waiting_hook);
+    _running_hook.swap_nodes(other._running_hook);
+    _translator = std::move(other._translator);
+    _start_time = other._start_time;
+    _total_waiting_time = other._total_waiting_time;
+    _total_running_time = other._total_running_time;
+    _slices_scheduled = other._slices_scheduled;
+    _current_wait_begin_time = other._current_wait_begin_time;
+    _current_running_begin_time = other._current_running_begin_time;
+    _stop_in_progress = other._stop_in_progress;
+    return *this;
+}
+
+translation_status translator_executable::status() const {
+    return _translator->status();
+}
+
+std::ostream& operator<<(std::ostream& os, const translator_executable& state) {
+    fmt::print(
+      os,
+      "{{translator: {}, time_since_start: {}, total_wait_time: {}, "
+      "total_running_time: {},  slices_scheduled: {}, stop_in_progress: {}, "
+      "running: {}, waiting: {}}}",
```

Comment:
nit: what's the rationale for logging the total wait/running times instead of the current times? I would've thought the current times are more helpful for logging/debugging. Would it make sense to log both?